pyteal/ast/scratchvar_test.py

```diff
@@ -66,3 +67,24 @@ def test_scratchvar_load():
     actual = TealBlock.NormalizeBlocks(actual)
 
     assert actual == expected
+
+def test_scratchvar_double_assign_invalid():
+    myvar    = ScratchVar(TealType.uint64, 10)
+    otherVar = ScratchVar(TealType.uint64, 10)
+    prog  = Seq([
+                myvar.store(Int(5)),
+                otherVar.store(Int(0))
+            ])
+    with pytest.raises(TealInternalError):
+        compileTeal(prog, mode=Mode.Signature, version=4) 
+
+    # Users cannot assign a slot that has been automatically
+    # assigned by the compiler
```

Comment:
In the current implementation, users cannot assign a slot ID that has been automatically assigned by the compiler. This is a little tricky as users have to ultimately keep track of the compiler assigned slots (which are deterministically assigned in ascending order, starting from 0, for a valid program) if they also want user defined slot IDs. 

Maybe there should be a stricter requirement for users to specify all the slot IDs or let the compiler automatically assign all of them. Mixing the two could _potentially_ be a source for bugs?

Looking forward, optimizing for "dead" slots may also affect this?

Reply:
I think I did not fully explain the intended behavior when mixing compiler-assigned slots and user-defined ones.

What should happen in this situation is that the compiler will first "reserve" all of the user-defined slot IDs (and throw an error if multiple slots request the same user-defined ID). Then, for the compiler-assigned slots, it can arbitrarily assign slots IDs from the range 0-255, excluding any reserved IDs that were already assigned to user-defined slots. (Ideally the compiler-assigned IDs will be backwards compatible with the current code to assign IDs, so that the generated TEAL code won't change after this update.)

Also, sorry for the confusing naming here, but the `id` field is not actually the TEAL slot ID that the `ScratchSlot` will have in the compiled program. It's just supposed to be a unique way to identify and deterministically order different `ScratchSlot` objects. If you look at the `compileTeal` function (specifically the loop `for index, slot in enumerate(sorted(slots, key=lambda slot: slot.id))`), you'll see that the `.id` property isn't actually what's assigned to the compiled source.

If you want to rename the `id` field to make it less confusing, I would support that.

Reply:
I think I understand now. To double check:

* The user is specifying a slot in the **compiled source**, so instead of assigning slots by `index, slot`, the compiler needs to assign the user-specified slots, then assign any extra ones in between (starting from 0 up to 255). 
* Basically there is a mapping between the `id` field and the actual slot number.

Reply:
Yes, that's correct